README.md
# Bloonix Plugin Configuration Files

## About Bloonix

For general information about Bloonix, questions, ratings or other things please visit our base repository:

* https://github.com/bloonix/bloonix

## Bloonix Plugins

Ordered by id (desc)

      id   |             plugin              |          command          
    -------+---------------------------------+---------------------------
       200 | Bloonix.Server.Check            | check-bloonix-server
        79 | Linux.Java8.Jstat               | check-jstat8
        78 | Linux.Java.Jstack               | check-jstack
        77 | Linux.Java7.Jstat               | check-jstat7
        76 | CMS.Typo3.Check                 | check-typo3-updates
        75 | Linux.NF.Conntrack.Check        | check-nf-conntrack
        74 | Linux.Gluster.Status            | check-gluster-status
        73 | Linux.Mount.Check               | check-findmnt
        72 | HTTP.Perf.Check                 | check-http-perf
        71 | NTP.Time.Check                  | check-ntp-time
        70 | RBL.Check                       | check-rbl
        69 | SNMP.Walk.Check                 | check-snmp-walk
        68 | SNMP.Check                      | check-snmp
        67 | SNMP.ServiceCheck               | check-snmp-service
        66 | SNMP.ProcessCheck               | check-snmp-process
        65 | SNMP.NumberOfProcesses          | check-snmp-nprocs
        64 | SNMP.Memory                     | check-snmp-mem
        63 | SNMP.Disk                       | check-snmp-disk
        62 | SNMP.CPU                        | check-snmp-cpu
        61 | Linux.RAID.LSI.Check            | check-lsi-raid
        60 | Bloonix.Satellite.Check         | check-bloonix-satellite
        59 | Simple.Wrapper                  | check-simple-wrapper
        58 | WebTransaction                  | check-wtrm
        51 | Varnish.Check                   | check-varnish3
        50 | Linux.Sensors                   | check-lm-sensors
        49 | Redis.Queue.Check               | check-redis-queue
        48 | Redis.Check                     | check-redis
        47 | Postfix.Mailqueue               | check-postfix-mailqueue
        46 | PostgreSQL.ProcessStatus        | check-pgsql8-procs
        45 | PostgreSQL.DatabaseStatus       | check-pgsql8-db
        44 | Nginx.Check                     | check-nginx
        43 | MySQL.SlaveStatus               | check-mysql-slave
        42 | MySQL5.Status                   | check-mysql5
        41 | Memcached.Check                 | check-memcached
        40 | Linux.Socket.Check              | check-sockstat
        39 | Linux.Smart.Health              | check-smart-health
        38 | Linux.Service.Check             | check-service
        36 | Linux.Processes.Check           | check-procstat
        35 | Ping.Check                      | check-ping
        34 | Linux.Paging.Check              | check-pgswstat
        33 | Linux.OpenFiles.Check           | check-open-files
        32 | Linux.NFS4Client.Check          | check-nfs4-client
        31 | Linux.NFS4Server.Check          | check-nfs4
        30 | Linux.NFS3.Check                | check-nfs3
        29 | Linux.NetstatPort.Check         | check-netstat-port
        28 | Linux.Netstat.Check             | check-netstat
        27 | Network.MTR.Check               | check-mtr
        26 | Linux.Memory.Check              | check-memstat
        25 | Linux.MDADM.Check               | check-mdadm
        24 | Linux.LoadAVG.Check             | check-loadavg
        23 | Linux.Updates.Check             | check-linux-updates
        22 | Linux.DiskIO.Check              | check-iostat
        21 | Linux.Interface.Statistics      | check-ifstat
        20 | Linux.Interface.Check           | check-iflink
        19 | Linux.Disk.Check                | check-diskusage
        18 | Linux.CPU.Check                 | check-cpustat
        17 | Linux.Bonding.Check             | check-bonding
        16 | Lighttpd.Status                 | check-lighttpd
        15 | Linux.DRBD.Check                | check-drbd
        14 | UDP.Check                       | check-udp
        13 | TCP.Check                       | check-tcp
        12 | SNMP.Interface                  | check-snmp-if
        11 | SMTP.Check                      | check-smtp
        10 | POP3.Check                      | check-pop3
         9 | Logfile.Check                   | check-logfile
         8 | IMAP.Check                      | check-imap
         7 | HTTP.Check                      | check-http
         6 | FTP.Check                       | check-ftp
         5 | Filestat.Check                  | check-filestat
         4 | DNS.Check                       | check-dns
         3 | Database.Check                  | check-dbconnect
         2 | Cluster.ServiceCheck            | check-cluster
         1 | Apache2.Statistics              | check-apache2



ChangeLog
bloonix-plugin-config (0.50) released at 2018-01-22

    * New plugins: check-jstat7, check-jstat8, check-jstack

bloonix-plugin-config (0.49) released at 2017-10-17

    * check-typo3-updates: Added new plugin.

bloonix-plugin-config (0.48) released at 2017-08-10

    * check-ping: ipv6 is now possible to use without --use-ipv6.

bloonix-plugin-config (0.47) released at 2017-06-28

    * check-wtrm: added param --ssl-protocol.

bloonix-plugin-config (0.46) released at 2017-06-22

    * check-nfs3: different new stats.

bloonix-plugin-config (0.45) released at 2017-06-09

    * check-http: added --http-header.

bloonix-plugin-config (0.44) released at 2017-01-25

    * Fixed stacked area charts with negative values.

bloonix-plugin-config (0.43) released at 2016-12-20

    * check-nf-conntrack: improved chart for count/max

bloonix-plugin-config (0.42) released at 2016-12-20

    * check-nf-conntrack: missed key 'free'

bloonix-plugin-config (0.41) released at 2016-12-20

    * New plugin: check-nf-conntrack

bloonix-plugin-config (0.40) released at 2016-12-12

    * check-iostat updated.

bloonix-plugin-config (0.39) released at 2016-09-14

    * Improved options of check-gluster-status.

bloonix-plugin-config (0.38) released at 2016-09-13

    * Added plugin: check-findmnt
    * Added plugin: check-gluster-status

bloonix-plugin-config (0.37) released at 2016-09-09

    * Added support for NTLM in check-smtp.

bloonix-plugin-config (0.36) released at 2016-08-12

    * New chart for inode usage.

bloonix-plugin-config (0.35) released at 2016-08-12

    * Kicked plugin-http-perf.

bloonix-plugin-config (0.34) released at 2016-08-10

    * Fixed colors of plugin-nfs4.

bloonix-plugin-config (0.33) released at 2016-03-28

    * Extra release because the gpg key of bloonix is updated.

bloonix-plugin-config (0.32) released at 2016-03-19

    * New plugin check-proc-status.

bloonix-plugin-config (0.31) released at 2016-03-02

    * Plugin check-logfile: permit non existent logfile.

bloonix-plugin-config (0.30) released at 2016-01-07

    * Plugin check-mysql-slave: new parameter verbose-error.

bloonix-plugin-config (0.29) released at 2015-12-29

    * Plugin check-memstat: new key memavailable and a new chart.

bloonix-plugin-config (0.28) released at 2015-12-16

    * Just changed the option order of check-http.

bloonix-plugin-config (0.27) released at 2015-11-30

    * New desc for plugin check-wtrm.

bloonix-plugin-config (0.26) released at 2015-11-26

    * New param per-cpu for plugin check-loadavg.

bloonix-plugin-config (0.25) released at 2015-11-20

    * New param authkey for plugin check-by-satellite.

bloonix-plugin-config (0.24) released at 2015-09-28

    * Infos updates for plugin check-linux-updates.

bloonix-plugin-config (0.23) released at 2015-09-15

    * New plugin check-bloonix-satellite.
    * New plugin check-ntp-time.

bloonix-plugin-config (0.22) released at 2015-08-15

    * Updated example of check-lm-sensors.

bloonix-plugin-config (0.21) released at 2015-06-16

    * Fixed description of cpu field "nice"

bloonix-plugin-config (0.20) released at 2015-06-16

    * New key "other" in bloonix-plugins-linux/check-cpustat.

bloonix-plugin-config (0.19) released at 2015-05-07

    * Set string "{}" as default for column "info" in table "plugin".
    * Plugin check-bloonix-server added.

bloonix-plugin-config (0.18) released at 2015-04-22

    * Different plugin changes.

bloonix-plugin-config (0.17) released at 2015-04-06

    * All snmp plugins has new options.

bloonix-plugin-config (0.16) released at 2015-03-09

    * Some new options. check-mdadm has a new parameter.

bloonix-plugin-config (0.15) released at 2015-01-26

    * Delete logger section in bloonix-load-plugins to prevent
      the creation of the file bloonix-server.log as user root.

bloonix-plugin-config (0.14) released at 2015-01-13

    * Fixed datatype for plugin-rbl.

bloonix-plugin-config (0.13) released at 2015-01-13

    * New plugin check-rbl.

bloonix-plugin-config (0.12) released at 2014-12-23

    * Replaced nagios-wrapper with simple-wrapper.

bloonix-plugin-config (0.11) released at 2014-12-04

    * New plugin check-snmp-walk.

bloonix-plugin-config (0.10) released at 2014-12-02

    * New plugin check-snmp.

bloonix-plugin-config (0.9) released at 2014-11-30

    * Different updates for linux and postfix plugins.

bloonix-plugin-config (0.8) released at 2014-11-28

    * Kicked plugins for Windows.

bloonix-plugin-config (0.7) released at 2014-11-26

    * Updates for bloonix-wtrm.

bloonix-plugin-config (0.6) released at 2014-11-25

    * Added plugin plugin-lsi-raid.
    * Added snmp plugins.

bloonix-plugin-config (0.5) released at 2014-11-08

    * Kicked deprecated base plugins.

bloonix-plugin-config (0.4) released at 2014-11-06

    * Added PluginLoader.pm and bloonix-load-plugins to
      load the plugins into the database.

bloonix-plugin-config (0.3) released at 2014-11-05

    * Update of check-linux-updates, check-smart-health

bloonix-plugin-config (0.2) released at 2014-11-03

    * Updated the license information.
    * Different plugin updates.

bloonix-plugin-config (0.1) released at 2014-09-10

    * Initial release.


Configure.PL
#!/usr/bin/perl
use strict;
use warnings;
use Getopt::Long qw(:config no_ignore_case);

my $prefix = "/usr/local";
my $o_confdir = "/etc";
my $o_perl = "/usr/bin/perl";
my $o_perl_prefix = "/usr/local";
my $o_perl_installdirs = "site";
my $o_without_perl = 0;
my $help = undef;

if (!-e "Makefile") {
    print "Execute me from the source directory!\n";
    exit 1;
}

GetOptions(
    "prefix=s" => \$prefix,
    "confdir=s" => \$o_confdir,
    "perl=s" => \$o_perl,
    "perl-prefix=s" => \$o_perl_prefix,
    "perl-installdirs=s" => \$o_perl_installdirs,
    "without-perl" => \$o_without_perl,
    "h|help" => \$help
);

if ($help) {
    print "Usage: perl Configure.PL [ OPTIONS ]\n\n";
    print "Configure options:\n\n";
    print "  --prefix DIR\n";
    print "  --confdir DIR (default: $o_confdir)\n";
    print "  --perl PERL EXECUTABLE (default: $o_perl)\n";
    print "  --perl-prefix DIR (default: $o_perl_prefix)\n";
    print "  --perl-installdirs INSTALLDIRS (default: $o_perl_installdirs)\n";
    print "  --help (print the help)\n\n";
    exit 0;
}

open my $fh, ">", "Makefile.config" or die "Unable to overwrite Makefile - $!";

print "Writing Makefile.config ...\n";

print $fh <<EOF;
# Base paths
PREFIX=\$(DESTDIR)$prefix
CONFDIR=\$(DESTDIR)$o_confdir
PERL=$o_perl
PERL_PREFIX=\$(DESTDIR)$o_perl_prefix
PERL_INSTALLDIRS=$o_perl_installdirs
WITHOUT_PERL=$o_without_perl
EOF

close $fh;

print "Done.\n";


perl/lib/Bloonix/PluginLoader.pm
package Bloonix::PluginLoader;

=head1 NAME

Bloonix::PluginLoader - Import bloonix plugins.

=head1 SYNOPSIS

    Bloonix::PluginLoader->run
        config_file => $config,
        plugin_file => $plugin,
    );

=head1 DESCRIPTION

Import a plugin into the bloonix database.

=head1 COPYRIGHT

Copyright (C) 2009-2014 by Jonny Schulz. All rights reserved.

=cut

use strict;
use warnings;
use JSON;
use Bloonix::DBI;
use Bloonix::Config;
use Params::Validate qw();
use base qw(Bloonix::Accessor);

__PACKAGE__->mk_accessors(qw/dbi sql json/);

our $VERSION = "0.50";

sub run {
    my ($class, %argv) = @_;
    my $self = bless \%argv, $class;
    my @plugins;

    if ($argv{plugin_file}) {
        push @plugins, split /,/, $argv{plugin_file};
    } elsif ($argv{plugin_path}) {
        $argv{plugin_path} =~ s!/\s*\z!!;
        push @plugins, <$argv{plugin_path}/plugin-*>;
        chomp @plugins;
    } elsif ($argv{load_all}) {
        foreach my $dir ("/usr", "/usr/local") {
            my $plugin_path = "$dir/lib/bloonix/etc/plugins/import";
            if (-d $plugin_path) {
                foreach my $dir ($self->get_files($plugin_path)) {
                    if (-d "$plugin_path/$dir") {
                        foreach my $plugin ($self->get_files("$plugin_path/$dir")) {
                            if (-f "$plugin_path/$dir/$plugin" && $plugin =~ /^plugin-/) {
                                push @plugins, "$plugin_path/$dir/$plugin";
                            }
                        }
                    }
                }
            }
        }
    }

    if (!@plugins) {
        die "no plugin-* files found\n";
    }

    my $config = Bloonix::Config->parse($self->{config_file});
    delete $config->{database}->{logger};
    $self->{config} = $config;
    $self->{dbi} = Bloonix::DBI->new($config->{database});
    $self->{sql} = $self->dbi->sql;
    $self->{json} = JSON->new();

    $self->dbi->connect;

    foreach my $plugin (@plugins) {
        eval{
            print "Load plugin $plugin\n";
            open my $fh, "<", $plugin
                or die "unable to open plugin '$plugin' for reading: $!";
            my $plugin_config = do { local $/; <$fh> };
            close $fh;
            $plugin_config = JSON->new->decode($plugin_config);
            if (ref $plugin_config->{plugin}->{info} eq "HASH") {
                $plugin_config->{plugin}->{info} = JSON->new->encode($plugin_config->{plugin}->{info});
            } else {
                $plugin_config->{plugin}->{info} = "{}";
            }
            $self->{plugin} = $self->validate($plugin_config);
            $self->update_plugin;
            $self->update_statistics;
            $self->update_charts;
        };
        warn $@ if $@;
    }
}

sub get_files {
    my ($self, $path) = @_;

    opendir my $dh, $path
        or die "unable to open directory '$path' for reading: $!";
    my @files = grep /^[a-z]/, readdir $dh;
    closedir $dh;

    return @files;
}

sub update_plugin {
    my $self = shift;
    my $data = $self->{plugin}->{plugin};

    my $plugin = $self->dbi->unique(
        $self->sql->select(
            table  => "plugin",
            column => "*",
            condition => [ id => $data->{id} ]
        )
    );

    if ($plugin) {
        $self->dbi->do(
            $self->sql->update(
                table  => "plugin",
                column => $data,
                condition => [ id => $data->{id} ],
            )
        );
    } else {
        $self->dbi->do(
            $self->sql->insert(
                table  => "plugin",
                column => $data,
            )
        );
    }
}

sub update_statistics {
    my $self = shift;
    my $plugin = $self->{plugin}->{plugin};
    my $data = $self->{plugin}->{statistic};
    my %stats = ();

    my $temp = $self->dbi->fetch(
        $self->sql->select(
            table => "plugin_stats",
            column => "*",
            condition => [ plugin_id => $plugin->{id} ]
        )
    );

    foreach my $stat (@$temp) {
        my $key = $stat->{statkey};

        if (exists $data->{$key}) {
            $stats{$key} = $stat;
        } else {
            $self->dbi->do(
                $self->sql->delete(
                    table => "plugin_stats",
                    condition => [
                        plugin_id => $plugin->{id},
                        statkey => $key
                    ]
                )
            );
        }
    }

    foreach my $key (keys %$data) {
        if (exists $stats{$key}) {
            $self->dbi->do(
                $self->sql->update(
                    table  => "plugin_stats",
                    column => $data->{$key},
                    condition => [
                        plugin_id => $plugin->{id},
                        statkey => $key
                    ]
                )
            );
        } else {
            $self->dbi->do(
                $self->sql->insert(
                    table => "plugin_stats",
                    column => $data->{$key},
                )
            );
        }
    }
}

sub update_charts {
    my $self = shift;
    my $plugin = $self->{plugin}->{plugin};
    my $data = $self->{plugin}->{chart};
    my %charts = ();

    foreach my $id (keys %$data) {
        my $chart = $self->dbi->unique(
            $self->sql->select(
                table => "chart",
                column => "*",
                condition => [ id => $id ]
            )
        );

        if (ref $data->{$id}->{options}->{series} eq "HASH") {
            $data->{$id}->{options}->{series} = [ $data->{$id}->{options}->{series} ];
        }

        $data->{$id}->{options} = $self->json->encode($data->{$id}->{options});

        if ($chart) {
            $self->dbi->do(
                $self->sql->update(
                    table  => "chart",
                    column => $data->{$id},
                    condition => [ id => $id ]
                )
            );
        } else {
            $self->dbi->do(
                $self->sql->insert(
                    table  => "chart",
                    column => $data->{$id}
                )
            );
        }
    }
}

sub validate {
    my $self = shift;

    my %options = Params::Validate::validate(@_, {
        plugin => {
            type => Params::Validate::HASHREF,
        },
        statistic => {
            type => Params::Validate::HASHREF | Params::Validate::ARRAYREF,
            default => [ ],
        },
        chart => {
            type => Params::Validate::HASHREF | Params::Validate::ARRAYREF,
            default => [ ],
        },
    });

    $options{plugin} = $self->validate_plugin($options{plugin});

    my $stats = ref $options{statistic} eq "HASH"
        ? [ $options{statistic} ]
        : $options{statistic};

    my $charts = ref $options{chart} eq "HASH"
        ? [ $options{chart} ]
        : $options{chart};

    $options{statistic} = { };
    $options{chart} = { };

    foreach my $stat (@$stats) {
        my $validated = $self->validate_statistic($stat);
        $validated->{plugin_id} = $options{plugin}{id};
        if (!$validated->{stattype}) {
            if ($validated->{datatype} =~ /varchar/) {
                $validated->{stattype} = "string";
            } else {
                $validated->{stattype} = "gauge";
            }
        }
        my $statkey = $validated->{statkey};
        $options{statistic}{$statkey} = $validated;
    }

    foreach my $chart (@$charts) {
        my $validated = $self->validate_chart($chart);
        $validated->{id} = sprintf("%d%04d", $options{plugin}{id}, $validated->{id});
        $validated->{plugin_id} = $options{plugin}{id};
        my $id = $validated->{id};
        $options{chart}{$id} = $validated;
    }

    return \%options;
}

sub validate_plugin {
    my $self = shift;

    my %options = Params::Validate::validate(@_, {
        id => {
            type => Params::Validate::SCALAR,
            regex => qr/^[1-9]\d*\z/
        },
        plugin => {
            type => Params::Validate::SCALAR,
            regex => qr/^[A-Za-z_0-9\-\.]+\z/
        },
        command => {
            type => Params::Validate::SCALAR,
            regex => qr/^check-[a-z0-9\-]{1,30}(\.[a-zA-Z]+){0,1}\z/
        },
        datatype => {
            type => Params::Validate::SCALAR,
            regex => qr/^(?:statistic|table|logfile|none)\z/
        },
        category => {
            type => Params::Validate::SCALAR,
            regex => qr/^[A-Za-z0-9,]{3,100}\z/
        },
        description => {
            type => Params::Validate::SCALAR,
            regex => qr/^[^<>]{3,500}\z/
        },
        abstract => {
            type => Params::Validate::SCALAR,
            regex => qr/^[^<>]{3,100}\z/
        },
        subkey => {
            type => Params::Validate::SCALAR,
            regex => qr/^\w{1,20}\z/,
            default => 0
        },
        company_id => {
            type => Params::Validate::SCALAR,
            regex => qr/^\d+\z/,
            default => 1
        },
        netaccess => {
            type => Params::Validate::SCALAR,
            regex => qr/^(yes|no)\z/,
            default => "no"
        },
        prefer => {
            type => Params::Validate::SCALAR,
            regex => qr/^(localhost|localnet|remote)\z/,
            default => "localhost"
        },
        worldwide => {
            type => Params::Validate::SCALAR,
            regex => qr/^(yes|no)\z/,
            default => "no"
        },
        info => {
            type => Params::Validate::SCALAR
        }
    });

    $options{netaccess} = $options{netaccess} eq "yes" ? 1 : 0;
    $options{worldwide} = $options{worldwide} eq "yes" ? 1 : 0;

    return \%options;
}

sub validate_statistic {
    my $self = shift;

    my %options = Params::Validate::validate(@_, {
        statkey => {
            type => Params::Validate::SCALAR,
            regex => qr/^\w{1,25}\z/
        },
        stattype => {
            type => Params::Validate::SCALAR,
            regex => qr/^(?:gauge|counter|string)\z/,
            optional => 1
        },
        datatype => {
            type => Params::Validate::SCALAR,
            regex => qr/^(?:smallint|integer|bigint|float|varchar\(\d+\))\z/
        },
        description => {
            type => Params::Validate::SCALAR,
            regex => qr/^[^<>]{3,500}\z/
        },
        substr => {
            type => Params::Validate::SCALAR,
            regex => qr/^\d+\z/,
            default => 0
        },
        units => {
            type => Params::Validate::SCALAR,
            regex => qr/^((kilo|mega|giga|tera|peta|exa|zetta|yotta){0,1}(bytes|bits)|ms|percent|null|temperature|unixtime|default)+\z/,
            default => "default"
        },
        alias => {
            type => Params::Validate::SCALAR,
            regex => qr/^.{0,100}\z/,
            optional => 1
        }
    });

    return \%options;
}

sub validate_chart {
    my $self = shift;

    my %options = Params::Validate::validate(@_, {
        id => {
            type => Params::Validate::SCALAR,
            regex => qr/^[1-9]\d{0,2}\z/
        },
        title => {
            type => Params::Validate::SCALAR,
            regex => qr/^[^<>]{3,50}\z/
        },
        options => {
            type => Params::Validate::HASHREF
        }
    });

    return \%options;
}

1;
